### ISSUE TITLE
Freeze version of hlint used to lint ghcide

### ISSUE THREAD

**Issue by pepeiborra** [haskell/haskell-language-server#2143]
Every time that @ndmitchell releases a new hint, our CI becomes blocked (the most recent one is https://github.com/haskell/haskell-language-server/pull/2116#issuecomment-907694079)

To prevent this, we should change the lint script to stick to a fixed version of hlint. This will probably imply forking Neil's script:

https://github.com/haskell/haskell-language-server/blob/a800b9d71e4eac70d672d30b2488949ea26dcbc3/fmt.sh#L1-L3

**Comment by ndmitchell** [comment]
The script ultimately calls on to https://github.com/ndmitchell/neil/blob/master/misc/run.sh. If you want to pin a version I think the right thing to do is read an optional external variable like `PINNED_VERSION` or similar and incorporate that on https://github.com/ndmitchell/neil/blob/master/misc/run.sh#L33.

I'd be happy to add such a feature.

**Comment by fendor** [comment]
Why not refer to an exact commit? Seems simpler than adding an environment variable

**Comment by ndmitchell** [comment]
If you want a precompiled binary (which you probably do) then you want to be pointing at a binary release. Those are by version number. Note that the script itself doesn't change, it's just driven by the GitHub releases, and currently always grabs the latest.

**Comment by jneira** [comment]
The ci is throwing ocasional errors trying to get the hlint script (https://github.com/jneira/haskell-language-server/runs/3532327041?check_suite_focus=true):

```
Run ./fmt.sh
  ./fmt.sh
  shell: C:\Program Files\Git\bin\bash.EXE --noprofile --norc -e -o pipefail {0}
Downloading and running hlint...
curl: (6) Could not resolve host: github.com
Error: Process completed with exit code 6.
```

So i wonder if we should host directly it to make it more reliable

**Comment by Anton-Latukha** [comment]
Also, there is a possibility to use GitHub action for it.
[Example setup](https://github.com/haskell-nix/hnix/blob/master/.github/workflows/HLint.yml).

Benefits would be:
  * HLint installs once, [not 20 times](https://github.com/haskell/haskell-language-server/actions/runs/1256370799]
  * HLint runs once, [not 20 times](https://github.com/haskell/haskell-language-server/actions/runs/1256370799)
    
    Running HLint once for reference GHC version seems fine and easy mentally. Sometimes HLint rules can apply to new GHCs and not apply to old GHC CPP code & vise versa, which can often put developers between a rock and a hard place. Standardizing HLint run, for example, on one version of `base` & GHC - ensures everyone knows what HLint check requires to pass it, and check in itself becomes consistent.
    
    * As seen in the above example - action allows pinning the version (afaik git commit hash is also allowed to be used).
    * Results of the scan - appear in the PR review as:
    * It allows to check different paths, and set the exception level for them: just output results into build log, post as info level notes in file change review screen, post info as info & warnings as warnings, or raise all to error & status return.
      
      For example, if to raise everything to error it would be:
      (because use HLint during dev & pass HLint CI test before merge, - there is no currently pull requests to show it actively, but for example in logs:
      
      ![Screenshot-2021-09-21-12:24:37](https://user-images.githubusercontent.com/20933385/134146651-b7efc5e3-f6da-4594-80e5-231cd71e4327.png)
      I use `relude` and tuned custom rules at that moment. `maybeToMonoid` rule is one I was on the verge about.
      
      About appearance in the review process, for example, if to raise everything:
      ![Screenshot-2021-09-21-12:18:21](https://user-images.githubusercontent.com/20933385/134145852-5b554054-7d47-4aec-bdb0-695db5db05f8.png)
      
      As you can imagine - that allows solving all that by itself, the developer sees checks, solves them in Haskell code, or tweaks the HLint settings. At that point - only HLint automation rules are what needs to be managed, rules maintenance exponentially settles into almost zero, while the benefit of them is big. Of course, the management of HLint rules and strong constraints by them can spawn a counter-action holy war, I am lucky to be the sole maintainer of a project and a supporter of almost all rules HLint has so for me it was easy to settle quickly into [the list](https://github.com/haskell-nix/hnix/blob/master/.hlint.yaml), merging everything I respect and adding a number rules to leverage project (`Utils`) modification & extension of the prelude.
      
  * Action also allows having different configs for different projects/directories.
  * There are several (at least two) GitHub actions for HLint, and they are comparable in features, I chose the one mentioned in the CI.
  * There is also GitLab version: https://gitlab.com/pipeline-components/haskell-hlint
    
Downsides:
    * Existing actions are not official.
    * and not actively maintained, so there can be time when they would stop working. But their simplicity is such that they can be forked and reworked.
    * If the project is moving from GitHub - actions can not be moved, HLint setup & configs can, but CI action probably would not be isomorphed.

I currently do not know other downsides.

`neil` uses own tool through unpublished GitHub-action:
  * `neil` [CI](https://github.com/ndmitchell/neil/blob/6c5a2d5d5f5a5d8fde2de63794ab5da216aa4364/.github/workflows/ci.yml#L33)
  * `hlint` [CI](https://github.com/ndmitchell/hlint/blob/26e503b33dbf2989c68e6a410f7b1d56ba86e1a6/.github/workflows/ci.yml#L33)
So maybe that way of using it gives more stability.

But it is definitely good to run action once per CI set, as such 20-times-for-1-change frequently hit access/request quotas & GitHub, Docker, Travis, Nix CDN & such love to block those kinds of spammy/DDoS-alike requests.

**Comment by jneira** [comment]
@Anton-Latukha thanks for the detailed analysis, the use of a dedicated action looks great and maybe we could create a dedicated job in the test workflow to run hlint on only 3 ghcs and one so (linux) if we want more coverage per ghc. 

We had a new issue recently running hlint: https://github.com/haskell/haskell-language-server/pull/2214#issuecomment-923653451 and i created an issue upstream https://github.com/ndmitchell/neil/issues/55 but maybe it is due the anti-DDoS as you mention

What do you think about contributing in the repo changing the way we use hlint? It could follow my suggestion (add one job in test workflow per 3 major ghc versions, it should throw an error and no continue with the existing workflow) if you think it makes sense

**Comment by Anton-Latukha** [comment]
Neil is of course the best person to create an official HLint action.
But if for @ndmitchell it would be seen beneficial to some to do - I can apply an evening to create a version to be merged into the official Haskell actions, or into HLint, or just host one in my account. Either way, the GitHub Actions API and abilities had several updates after mentioned actions were created, API deprecations, the introduction of whole new formats as Yaml specifications comparing to JS/JSON ones, (non VM) fast containerized actions.

**Comment by Anton-Latukha** [comment]
@jneira

> What do you think about contributing in the repo changing the way we use hlint?

Yes, it is a thing I also wanted to propose & motivated to do. Depending on Mitchell response I can prioritize actions accordingly (do the HLS CI config change, or create an action and then introduce config change here).

As you proposed it also the second part (submit CI config restructure) is agreed. Upon configuration work going to reach out in the nearest days in IRC/Matrix, to make it under maintainer team requirements.

**Comment by Anton-Latukha** [comment]
The https://github.com/haskell/haskell-language-server/pull/2537 solves the topic. There I am pinning to `3.3.1`.

**Comment by ndmitchell** [comment]
This looks like it is solved. @Anton-Latukha has switched to a GitHub actions plugin, which is being attempted to be merged with the Haskell github action, so that all seems like we're sorted now.

**Comment by jneira** [comment]
Solved by #2357
